### ISSUE TITLE
Rust: Experiment with Profiles

### ISSUE THREAD

**Issue by kazk** [codewars/runner#101]
[Profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#profiles) are used to set compiler options, and we're currently using the default [dev profile](https://doc.rust-lang.org/cargo/reference/profiles.html#dev) (the test executable uses the test profile):

```toml
[profile.dev]
opt-level = 0
debug = true
debug-assertions = true
overflow-checks = true
lto = false
panic = 'unwind'
incremental = true
codegen-units = 256
rpath = false
```

Crates are also built using that, but I found that we can be optimize them independently by [overriding](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides):
```toml
# Set the default for dependencies.
[profile.dev.package."*"]
opt-level = 2
```

We need to experiment with different profiles to find one that works well for our use case.

**Comment by kazk** [comment]
To compare with optimized C++ on Codewars for kata with performance requirements, I translated monadius's [N queens problem (with one mandatory queen position) - challenge version](https://www.codewars.com/kata/5985ea20695be6079e000003) solution to Rust.

I translated almost exactly and it finishes constantly in about 2-2.5 seconds (Rust is faster) :confused:

![image](https://user-images.githubusercontent.com/639336/105983308-6220d800-604d-11eb-9e12-c0932347ebc2.png)


<details>
<summary>Test Cases</summary>

```rust
#[cfg(test)]
mod tests {
    use rand::{
        prelude::{SliceRandom, ThreadRng},
        Rng,
    };

    use super::*;

    fn check_board(board: &[u8], n: usize, fixed: (usize, usize)) {
        let mut offset = 0;
        let mut num_queens = 0;
        let mut queens: Vec<Option<usize>> = vec![None; n];
        for y in 0..n {
            for x in 0..n {
                let field = board[offset];
                offset += 1;
                if field == b'Q' {
                    assert!(
                        queens[y].is_none(),
                        "The board should not have horizontal attacks between Queens"
                    );
                    num_queens += 1;
                    queens[y] = Some(x);
                } else if field != b'.' {
                    panic!("The board has invalid characters");
                }
            }
            assert_eq!(
                board[offset], b'\n',
                "The board has missing/incorrect characters"
            );
            offset += 1;
        }

        assert_eq!(num_queens, n, "The number of queens should be equal to n");

        let queens = queens.into_iter().map(Option::unwrap).collect::<Vec<_>>();
        assert!(
            queens[fixed.1] == fixed.0,
            "The fixed queen is not in the required position"
        );

        // Check no attacks
        let mut taken_cols = vec![false; n];
        let mut taken_diag1 = vec![false; 2 * n];
        let mut taken_diag2 = vec![false; 2 * n];
        for row in 0..n {
            let col = queens[row];
            assert!(
                !taken_cols[col],
                "The board has vertical attacks between Queens"
            );
            taken_cols[col] = true;
            assert!(
                !taken_diag1[col + row],
                "The board has diag1 attacks between Queens"
            );
            taken_diag1[col + row] = true;
            assert!(
                !taken_diag2[n + col - row - 1],
                "The board has diag2 attacks between Queens"
            );
            taken_diag2[n + col - row - 1] = true;
        }
    }

    fn test_solution(n: usize, fixed: (usize, usize)) {
        check_board(&solve_n_queens(n, fixed).as_bytes(), n, fixed);
    }

    fn test_range(rng: &mut ThreadRng, range: std::ops::Range<usize>) {
        let n = rng.gen_range(range);
        test_solution(n, (rng.gen_range(0..n), rng.gen_range(0..n)));
    }

    #[test]
    fn basic_tests() {
        let basic_tests = vec![(8, (3, 0)), (4, (2, 0)), (1, (0, 0))];
        for (n, fixed) in basic_tests.into_iter() {
            test_solution(n, fixed);
        }
    }

    #[test]
    fn no_solution_tests() {
        let no_solutions = vec![(2, (0, 0)), (3, (2, 0)), (6, (0, 0))];
        for (n, fixed) in no_solutions.into_iter() {
            assert_eq!(solve_n_queens(n, fixed), "", "No solution is possible");
        }
    }

    #[test]
    fn random_tests_small_n() {
        #[rustfmt::skip]
        const NO_SOLUTIONS: [(usize, (usize, usize)); 33] = [
            (2, (0, 0)), (2, (0, 1)), (2, (1, 0)), (2, (1, 1)),
            (3, (0, 0)), (3, (0, 1)), (3, (0, 2)),
            (3, (1, 0)), (3, (1, 1)), (3, (1, 2)),
            (3, (2, 0)), (3, (2, 1)), (3, (2, 2)),
            (4, (0, 0)), (4, (0, 3)),
            (4, (1, 1)), (4, (1, 2)),
            (4, (2, 1)), (4, (2, 2)),
            (4, (3, 0)), (4, (3, 3)),
            (6, (0, 0)), (6, (0, 5)),
            (6, (1, 1)), (6, (1, 4)),
            (6, (2, 2)), (6, (2, 3)),
            (6, (3, 2)), (6, (3, 3)),
            (6, (4, 1)), (6, (4, 4)),
            (6, (5, 0)), (6, (5, 5)),
        ];

        let mut rng = rand::thread_rng();
        let mut specs = vec![false; 5];
        specs.extend(vec![true; 5]);
        specs.shuffle(&mut rng);
        for has_solution in specs {
            if has_solution {
                loop {
                    let size = rng.gen_range(4..11);
                    let fixed = (rng.gen_range(0..size), rng.gen_range(0..size));
                    if !NO_SOLUTIONS.iter().any(|(n, f)| n == &size && f == &fixed) {
                        test_solution(size, fixed);
                        break;
                    }
                }
            } else {
                let (n, fixed) = NO_SOLUTIONS.choose(&mut rng).unwrap();
                assert_eq!(solve_n_queens(*n, *fixed), "", "No solution is possible");
            }
        }
    }

    #[test]
    fn random_tests_larger_n() {
        let mut rng = rand::thread_rng();
        test_range(&mut rng, 10..15);
        test_range(&mut rng, 15..20);
        test_range(&mut rng, 20..25);
        test_range(&mut rng, 25..35);
        test_range(&mut rng, 45..50);
    }

    #[test]
    fn random_tests_large_n() {
        let mut rng = rand::thread_rng();
        test_range(&mut rng, 100..150);
        test_range(&mut rng, 200..250);
        test_range(&mut rng, 300..350);
        test_range(&mut rng, 400..450);
        test_range(&mut rng, 450..500);
    }

    #[test]
    fn random_tests_even_larger_n() {
        let mut rng = rand::thread_rng();
        test_range(&mut rng, 500..600);
        test_range(&mut rng, 650..700);
        test_range(&mut rng, 750..800);
        test_range(&mut rng, 850..900);
        test_range(&mut rng, 950..1000);
    }
}
```

</details>

**Comment by kazk** [comment]
I thought the test cases were running in parallel or something. But combining into one test case doesn't change much. Maybe it takes slightly longer on average, like 2.2-2.6s, but it varies depending on the inputs to begin with and it's not significant.

![image](https://user-images.githubusercontent.com/639336/106048783-28c28980-609a-11eb-8c5b-4ed1d48b04ca.png)

---

I just published the [translation](https://www.codewars.com/kumite/6011e3d399cef9002b7c52d4). I double checked it's testing the same thing as C++, but Rust doesn't have any issue unoptimized.

**Comment by imjasonmiller** [comment]
Hi @kazk, 

Have the above changes already been implemented? I'm running into a related issue while translating the following kata to Rust:

* [URL shortener (5 kyu)](https://www.codewars.com/kata/5fee4559135609002c1a1841/javascript)

I have a translation drafted, but it's unable to do the required `475 000` iterations to test for all permutations. It times out at around `200 000` iterations. The issue seems to be with Rust running in `debug` mode with the usage of a large `HashMap`. 
 
I'm only focusing on `HashMap` here, so I've setup a basic example playground with the below code that highlights the difference in performance between `debug` and `release`:
 
 ```rust
use std::collections::HashMap;

fn main() {
    let now = std::time::Instant::now();
    
    // Create 475_000 urls
    let mut urls = HashMap::<i64, i64>::new();
    
    for i in 0..475_000 {
        urls.insert(i, i);
    }
    
    // Test against user implementation
    for (k, v) in urls.iter() {
        // Example test assertions
        assert!(k >= &0);
        assert!(v >= &0);
    }
    
    println!("Elapsed: {} ms", now.elapsed().as_millis());
}
```

Here are some rough timings for the above code from different runs

Mode | [Debug](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=852dd52b448edc0ac9c883bd7a7e0493) (ms) | [Release](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=092fffe71bf3d57d32bad75171772533) (ms)
---|---|---
Run 1 | 1894 | 56 
Run 2 | 1842 | 54
Run 3 | 1134 | 54
Run 4 | 1493 | 86
Run 5 | 1852 | 83
Mean | 1643 | 67

Locally running my drafted translation with different optimization levels for the tests themselves gave the following:

```toml
[profile.test]
opt-level = 3
```
Optimization level | 0 (ms) | 1 (ms) | 2 (ms) | 3 (ms)
---|---|---|---|---
Run 1 | 7272 | 5349 | 4281 | 3302 
Run 2 | 7289 | 5231 | 4321 | 3924
Run 3 | 7313 | 5236 | 4317 | 3302 
Run 4 | 7312 | 5221 | 4321 | 3321 
Run 5 | 7305 | 5240 | 4286 | 3288 
Mean | 7298 | 5255 | 4305 | 3427
 
While a large improvement, it didn't come close to `--release` yet, but I found that adding `opt-level = 2` for dependencies narrowed it down to  `--release` mode for my translation! A higher `opt-level` didn't seem to do much more in this case.

```toml 
[profile.dev.package."*"]
opt-level = 2
```

Optimization level | 0 (ms) | 1 (ms) | 2 (ms) | 3 (ms)
---|---|---|---|---
Run 1 | 3291 |1410 | 395 | 373 
Run 2 | 3307 | 1403 | 372 | 385
Run 3 | 3315 | 1406  | 383 | 382
Run 4 | 3355 | 1402 | 395 | 399
Run 5 | 3303 | 1410 | 371 | 371
Mean | 3314 | 1406 | 383 | 382

If you want to look at the drafted translation, I can send it via a private message.

**Comment by kazk** [comment]
> Have the above changes already been implemented?

No, what I posted above required no change.

The profile options (including `opt-level`) exists because there are trade-offs. The goal of this issue to understand them (especially the drawbacks) and explore what's best for Codewars. We need to experiment more with different kata/solutions to find what works for us. I'll close this issue when it's decided and changes are deployed if any.

Some questions are:

- What are the drawbacks of using a profile optimized for execution speed?
  - Slower feedback for tests not requiring performance. Is it significant enough to hurt user experience?
  - Generally takes longer to compile.
    - **Is it significant enough to prevent some challenges (involving macros, generics, futures, etc)?** Need some examples. This is important because unlike Haskell, we can't disable optimization when necessary. For execution speed, we can often work around by adjusting the tests, but we can't do that to reduce the time it takes to compile the optimized solution.
  - **Does it change anything else? Error messages? Warnings?** This is also important on Codewars where users cannot adjust these.
  - Not the default practice. This means we might run into some weird issues.
  - More likely to change between versions (the time required to compile optimized, the effect of the optimization, are not part of the stability guarantee). This can increase the effort required to update kata, which in turn makes it difficult to keep Rust up to date. I think keeping Rust relatively up to date is important because even though Rust guarantees stability, it's evolving fast and new features are constantly added.
  - Anything else?
- Is it possible to avoid some drawbacks using options in the profile? e.g., `debug-assertions` will be kept.
- Is there anything impossible to do with the current profile?
  - Is it possible to adjust the tests? The performance tests should have a specific goal and that shouldn't be just the number of times it can run.
  - Is there a better way to test performance requirements?
- Is the overall benefit significant enough to justify the drawbacks?

---


> The issue seems to be with Rust running in `debug` mode with the usage of a large `HashMap`.

If I remember correctly, Rust's std `HashMap` uses a hashing function optimized for security over speed. There are crates if you need something faster, and I'll consider adding them if necessary.

Some data structures are simply slower in `debug` because it contains a lot of code that's removed in the `release` mode for better developer experience (the number of lines in emitted assembly for the first minimal example are `44790 vs 1515`). Normally, it's not really noticeable because it's unlikely for the difference to accumulate as much outside Codewars.

---

> If you want to look at the drafted translation, I can send it via a private message.

Thanks, I'll take a look. Admins can see drafts :)

**Comment by kazk** [comment]
Found this on Reddit:

> You can tell Cargo to compile all your dependencies with full optimisation (like in Release mode), while compiling your own code with few optimisations (like in Development mode). Since your dependencies change rarely you can take the hit of compiling them in Release mode once without affecting your day-to-day workflow much. Potentially this can give you the best of both worlds - fast run time and fast(ish) compile times.
>
> ```
> [profile.dev.package."*"]
> # Set the default for dependencies in Development mode.
> opt-level = 3
>
> [profile.dev]
> # Turn on a small amount of optimisation in Development mode.
> opt-level = 1
> ```
>
> Of course the effectiveness will depend heavily on your own individual program, but worth trying I think.
>
> As the [Cargo profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides-and-generics) page shows, if you make heavy use of generics from 3rd party crates then you won't get the full benefit, but in most cases I would not be surprised if 90% of the Rust code in your program is actually from crates and can benefit from this. It certainly worked fantastically well on a game I was writing.
>
> [This is a neat trick for getting good runtime performance with Development mode builds](https://old.reddit.com/r/rust/comments/gvrgca/this_is_a_neat_trick_for_getting_good_runtime/)


---


> [Profiles - The Cargo Book](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides-and-generics)
>
> Further complicating the issue, `rustc` has some optimizations where it will attempt to share monomorphized generics between crates. If the opt-level is 2 or 3, then a crate will not use monomorphized generics from other crates, nor will it export locally defined monomorphized items to be shared with other crates. When experimenting with optimizing dependencies for development, consider trying opt-level 1, which will apply some optimizations while still allowing monomorphized items to be shared.

**Comment by kazk** [comment]
The performance of `HashMap` is a valid reason for the change. The heavy use of `HashMap` with `opt-level = 0` prevents some valid solutions and test setup and I couldn't find a workaround. For this example, using `fnv` hasher didn't help because the size of the key is not that small.

---

@imjasonmiller 

For your translation, using [`entry`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry) should make it a bit faster (`get` and `insert` does look up twice):

```diff
- if let Some(short_url) = urls.get(&rand_long_url) {
-     assert_eq!(*short_url, rand_short_url, "Should work with same URLs!");
- } else {
-     urls.insert(rand_long_url, rand_short_url);
- }
+ match urls.entry(rand_long_url) {
+     Entry::Occupied(o) => {
+         assert_eq!(*o.get(), rand_short_url, "Should work with same URLs!");
+     }
+     Entry::Vacant(v) => {
+         v.insert(rand_short_url);
+     }
+ }
```

---

Comparing combinations of `opt-level` for solution and tests using the translation's random tests with `475_000` URLs and above change:

| solution  | packages | duration |
| :-------: | :------: | -------: |
|         0 |        0 |     ~17s |
|         1 |        0 |     ~14s |
|         0 |        1 |     ~13s |
|         0 |        2 |      ~9s |
|         0 |        3 |      ~9s |
|         1 |        1 |      ~9s |
|         2 |        0 |      ~9s |
|         3 |        0 |      ~9s |
|         1 |        2 |      ~5s |
|         1 |        3 |      ~5s |
|         2 |        1 |      ~5s |
|         3 |        1 |      ~5s |
|         2 |        2 |    ~1.3s |
|         2 |        3 |    ~1.3s |
|         3 |        2 |    ~1.3s |
|         3 |        3 |    ~1.3s |

Note that this wasn't inside a container and the duration does not include compile time.

- `opt-level = 2` vs `opt-level = 3` for packages didn't make much difference for this example
- `opt-level` for packages shouldn't affect the time to compile solutions, so we can do `3`
- `opt-level = 1` for solution makes a significant difference (`0 3 ~9s` vs `1 3 ~5s`) and shouldn't make it too slow to compile (need to test this).
- `opt-level = 2` for solution makes a significant difference from `1` as well, but I'm concerned about the time to compile.

So, for now, the following looks good:
```toml
[profile.dev]
opt-level = 1
[profile.dev.package."*"]
opt-level = 3
[profile.test]
opt-level = 1
[profile.test.package."*"]
opt-level = 3
```

---

For this kata, I don't think it's necessary to test this many though. Why not measure duration with less URLs and assert it's lower than some number?

**Comment by imjasonmiller** [comment]
Yes! I asked around on Stackoverflow why _both_ the `test` and `dev` profile were lowering the total run time for a local version of my draft, but that's due to the `dev` profile also applying its `opt-level` to dependencies during `cargo test`, which can be made more specific with `dev.packages."*"` as you already noted. I was also linked the following post that mentions the same:

* [Stackoverflow, "Linking to optimised crate from a debug build"](https://stackoverflow.com/questions/31260406/linking-to-optimised-crate-from-a-debug-build/31271716#31271716)

--- 

I'll dig deeper into those questions and what possible drawbacks might be. I can answer one of them:

* **Is it possible to avoid some drawbacks using options in the profile? e.g., `debug-assertions` will be kept.**
  I think there's no harm in keeping this as is! I thought this option referred to [`std::assert`](https://doc.rust-lang.org/std/macro.assert.html), [`std::assert_eq!`](https://doc.rust-lang.org/std/macro.assert_eq.html) and friends at first, but it refers to [`std::debug_assert`](https://doc.rust-lang.org/std/macro.debug_assert.html), which are assertions that are optimized away when using the `--release` profile, as it has `debug-assertions` set to `false`. The assertions `assert`, `assert_eq!` which are used in tests are kept, even when compiling with `--release`.

--- 

If you're curious about some of the effects `opt-level` can have, I've setup a small local example that roughly highlights how much faster it runs by targeting external dependencies such as `rand`:

<details>
  <summary><b>cargo.toml</b></summary>

```toml
[package]
name = "profiles_test"
version = "0.1.0"
authors = ["John Doe <john@example.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rand = "0.8.3"

[profile.dev]
opt-level = 0

[profile.dev.package."*"]
opt-level = 2

[profile.test]
opt-level = 1
```
</details>

<details>
  <summary><b>main.rs</b></summary>

```rust
use rand::Rng;
use std::collections::HashMap;

fn expensive_fn() {
    let now = std::time::Instant::now();

    let mut map = HashMap::<u64, u64>::new();
    let mut rng = rand::thread_rng();

    for _ in 0..500_000 {
        map.insert(rng.gen(), rng.gen());
    }

    println!("{} ms", now.elapsed().as_millis());
}

fn main() {
    expensive_fn();
}

#[cfg(test)]
mod tests {
    use super::expensive_fn;

    #[test]
    fn profile() {
        expensive_fn();
    }
}
```
</details>

For the above code, I've taken the mean amount of milliseconds for 5 runs of `cargo test -- --nocapture`, resulting in the grid below:

&nbsp;                         | `test = 0` |  `test = 1` | `test = 2` | `test = 3`
-------------------------|-----------|------------|-----------|----------
`dev.packages."*" = 0` | 1660       | 1178         | 649         | 651 
`dev.packages."*" = 1` | 1263       |  709          | 273         | 264
`dev.packages."*" = 2` | 1098       |  499          | 69           | 68
`dev.packages."*" = 3` | 1104       |  489          | 62           | 68

I assume it's rare for common external dependencies such as `rand`, `regex` and others to produce serious errors, from which I reasoned that a more aggressive optimization level for these would perhaps be okay? Even with compiling `packages` at `2` and `test` at `1` there's already a significant jump in speed for this particular code using only `rand`.

---

> Thanks, I'll take a look. Admins can see drafts :)

Ha, that makes sense. I thought it'd be best not to post it here, as others might copy the solution.

--- 

> For this kata, I don't think it's necessary to test this many though. Why not measure duration with less URLs and assert it's lower than some number?

Yes, I think there could be alternative ways to test this. Perhaps changing the amount of available characters for URLs from 4, with 26^1 + 26^2 + 26^3 + 26^4 = 475254 permutations, to perhaps 3 characters with 18278 permutations? It is a significant reduction, but still challenges the user in finding an algorithm that allows for that many permutations in 3 lowercase letters.

I'll ask the kata author if that's an acceptable trade-off to them.

Have a great evening!

**Comment by kazk** [comment]
> Yes, I think there could be alternative ways to test this. Perhaps changing the amount of available characters for URLs from 4, with 26^1 + 26^2 + 26^3 + 26^4 = 475254 permutations, to perhaps 3 characters with 18278 permutations?

Oh, I didn't realize that the number came from allowed characters in the shortened URL. The description indirectly specifies that the id must have length `[1, 4]`. I guess that's fine then.

**Comment by kazk** [comment]
I tested against all published Rust kata and the change looked safe enough to me, so I decided to update Rust 1.49 in-place (was originally thinking of waiting until the next version).

I deployed to preview and your original translation with 475K URLs takes about ~9-10s total.

**Comment by imjasonmiller** [comment]
That's wonderful, looking forward to trying it out! Thank you for taking the time to do so :-) 

Have a great weekend.

**Comment by kazk** [comment]
Deployed. Please let me know if you notice anything weird.
Thank you for your help and all the Rust translations.
